### ISSUE TITLE
iOS Safari & Chrome cannot upload an avatar

### ISSUE THREAD

**Issue by Bekcpear** [flarum/core#1128]
###Expected behavior
Upload an avatar successful on iOS

###Actual behavior.
No response when I click the avatar upload button

###Specifications
- iOS Version: 10.2.1
- Version of Flarum: 0.1.0-beta.6
- Website URL where the bug is visible: https://pondof.fish
- The webserver you are running: nginx
- PHP version: 7.1.1
- Hosted environment: vps

iOS Safari & Chrome cannot upload an avatar too.
The difference I tested between "iOS Safari" and "MS Edge#1125" is "iOS Safari" cannot popup the file selection box while "MS Edge" cannot upload the selection image. 

I inspected the profile page on linux by iso-webkit-debug-proxy & mozilla/valence. No errors appear in WebIDE console.
Avatar can be uploaded successful if I unhide the INPUT element and click it manually.
Change `'<input type="file">'` to `'<input type="file" style="cursor: pointer">'` in `AvatarEditor.js` and rebuild the `forum-*.js` in `assets` folder is not working.

I also try the following code and these worked as expected on iOS (file selection box can popup when I click the "button" element) :
```
<!DOCTYPE html>
<html>
  <head>
    <script src="./jquery-2.1.4.js"></script>
  </head>
  <body>
    <script>
      $(function() {
        console.log("Ready!");
        $('<button id="1" style="width: 100px; height: 20px; float: left;"></button>').appendTo("body");
        $( "#1" ).click(function(){
            console.log("Click");                                                                                                                          
            $('<input type="file">').appendTo("body").hide().click();
        });
      })
    </script>
  </body>
</html>
```

**Comment by Kakifrucht** [comment]
Probably related to #1125

**Comment by chaosbunker** [comment]
I also have the same issue, with both chrome and safari in iOS

**Comment by oanhnn** [comment]
Yes. I have same issue and fixed it by changing upload avatar specs :smile:

**Comment by franzliedke** [comment]
@oanhnn Not sure I understand. How did you fix this exactly? Which specs are we talking about?

**Comment by oanhnn** [comment]
Spec:
Haven't avatar yet case: Click avatar => show dialog chose file => upload
Have avatar yet: Click avatar => show dialog chose change avatar or remove avatar => upload

My spec after fix:
Click avatar => show dialog chose file => upload

My solution: make input[file] in avatar block

**Comment by ibrahimk157** [comment]
@oanhnn I don't get it, what file are you editing?

**Comment by franzliedke** [comment]
To anybody who can reproduce this bug:
- Can you see any error message in the browser's dev tools console?
- In the dev tools' "Network" tab, do you see a request that tries to upload the file? If so, does it return successfully?
- Any errors in the server-side error logs?

**Comment by Bekcpear** [comment]
@franzliedke Thank you for your reply.

It's not a server problem.

It's just cannot popup the file dialog when I click the avatar upload button on iOS platform.

If I unhide the 'file input' element (which is triggered by JS before), and click the input manually, iOS file dialog can appears normally and then avatar uploaded successfully.

**Comment by franzliedke** [comment]
Okay, thanks for clarifying.

Apparently, we need to shift the file input field off-screen instead of [hiding it](https://github.com/flarum/core/blob/123c3a93f57cac94035acbcbaf4542183ec476c6/js/forum/src/components/AvatarEditor.js#L107).

I'll let @tobscure take care of this. ;)

**Comment by franzliedke** [comment]
Some references:
- https://github.com/fat/bean/issues/82
- https://stackoverflow.com/questions/25886480/trigger-click-on-input-file
- https://stackoverflow.com/questions/793014/jquery-trigger-file-input
- https://stackoverflow.com/questions/13290421/script5-access-is-denied-error-on-ie9-when-firing-click-from-onchange

**Comment by ibrahimk157** [comment]
@franzliedke flagrow/upload has a upload button that works all the time... can't you copy (or compare) the code from that to solve this problem? Or have you already discovered why this happens?

**Comment by franzliedke** [comment]
@ibrahimk157 [As far as I can tell](https://github.com/flagrow/upload/blob/06c66d5f09e0ef542a6620fb9fece65f4684d007/js/forum/src/components/UploadButton.js), they do not hide the upload input field the way we do.

**Comment by ibrahimk157** [comment]
@franzliedke so are you or toby able to fix it?

**Comment by franzliedke** [comment]
Using my suggestion above, I hope Toby will be able to fix it, yes. :)

**Comment by ibrahimk157** [comment]
@franzliedke Awesome! Thank you for all your hard work!

**Comment by franzliedke** [comment]
Apparently Flagrow [does hide the button](https://github.com/flagrow/upload/blob/a8405211e16b52092ce25bf55d73ee4ad63cb594/less/forum/upload.less#L1-L16), but not by setting its `display` to `none`, but rather by wrapping it with another CSS element and styling *that*. Might be even more inspiration for Toby. ;)

**Comment by WiseClock** [comment]
I came up with an extension that tweaks the avatar uploading function a little bit and was having the same issue at the beginning since I didn't change any of the code that's firing the click events.
Then today I did a little hack to it, now it works on iOS, too.

What causes this issue is that the current code in [AvatarEditor](https://github.com/flarum/core/blob/34f3d93ce56ad18ada6aad65bfc77244fea09716/js/forum/src/components/AvatarEditor.js) only adds a `click` handler to the elements (the `onclick` attributes).  In order to make it work on some mobile devices, we need to add the `tap` (or `touchend`) listener, too.  It'll be something like: `$("#id").on("tap click", function(e) { });`
But doing this sometimes triggers the event twice on mobile devices -- one for tap and one for click.  So at the end I manage to hack it by adding a `touchend` event without changing the old code, and inside the `touchend` event I `e.preventDefault();` to avoid it triggering the `click` event once more.  Then I can just simulate the mouse hover and manually trigger the `click` event again.  Hooray!

The codes I injected to make it happen are [here](https://github.com/WiseClock/flarum-ext-profile-image-crop/commit/35ef23cbb2f42156928f674fdf6525a3ac6cac26), note that I remove and add the event listeners depending on if the user has already an avatar or not.  It is because that the `config()` function is called every time when some elements are changed.  I need to remove the unused listeners or it will break ;)

**Comment by gingerbeardman** [comment]
+1 I see this issue on discuss.flarum.org

**Comment by TurboProgramming** [comment]
[Needs `mobile` and `UX` tags]

**Comment by datitisev** [comment]
I can no longer reproduce this on iOS 11.3.
I was able to upload an avatar with the expected delays of native iOS popups on that regard.

**Comment by ardacebi** [comment]
iOS 12 & Safari:

![e14decbc-695c-4c76-9676-d7b155550471](https://user-images.githubusercontent.com/17576065/44006239-9acb6b30-9e89-11e8-831d-92d03f54282d.jpeg)

**Comment by tobscure** [comment]
@ardacebi I think the issue you've encountered there ("oops something went wrong") might be unrelated to this issue ("No response when I click the avatar upload button"). Please verify.

**Comment by breathingboard** [comment]
This is still an issue. I’m on iOS 11.4.1, using Chrome version 68.0.3440.83, which I assume is the latest version as of today.

If I tap the plus sign just right, the iOS upload box slides up as it should. But, it only works every now and then. 

I think this is a fairly common problem with iOS and upload buttons if you aren’t using a standard upload button.

Again it’s very inconsistent. So you may think this is no longer an issue one day, then a week later try to upload and it doesn’t work.

**Comment by breathingboard** [comment]
Here's a potentially interesting tidbit. If you manage to get an avatar uploaded. Then, re-uploading works every time. Tapping on my existing account on my iPhone instantly pops up the upload or remove dialog. 

But, if I register as a new member, so that by default I just have the letter avatar, and I go to tap that to add my avatar for the first time, this is when the dialog box rarely appears for me on my iPhone. 

Maybe it's a CSS thing? I'm sure the CSS is different when you compare the default letter avatar with an uploaded avatar.

**Comment by johnlewissims** [comment]
Hello! Just wanted to follow up on this issue and see if anyone is experiencing this problem still.  On my site the issue occurs on all mobile browsers.  I've tested in Chrome, Safari and Brave. 

If i run it on desktop it works just fine. Which unfortunately means I don't have a console error to give anyone.

**Comment by gingerbeardman** [comment]
YI no longer use flare, but you can get console logs from iPhone using Mac Safari > Develop menu.

Quick google result: https://medium.com/@mattcroak718/debugging-your-iphone-mobile-web-app-using-safari-development-tools-71240657c487

**Comment by johnlewissims** [comment]
Update.  Just updated my phone to the latest iOS 13.4.1 (was running 13.3.1).  Everything works as expected. 

Update your phones folks.

**Comment by askvortsov1** [comment]
Just replicated on iPad, Google Chrome v87. The OS's file upload interface showed up once, then I couldn't get it to do so again.

Safari works fine though.

Why, Apple, must you be like this?

**Comment by zerosonesfun** [comment]
> Just replicated on iPad, Google Chrome v87. The OS's file upload interface showed up once, then I couldn't get it to do so again.
> 
> Safari works fine though.
> 
> Why, Apple, must you be like this?

For what it’s worth I’ve used Flarum on two different model iPhones with Safari and Chome. I haven’t experienced this. Yet.
